discord-bot/src/helpers.ts

```diff
@@ -0,0 +1,235 @@
+import { FullMarket } from 'common/api-market-types'
+import * as console from 'console'
+import {
+  EmbedBuilder,
+  Message,
+  MessageReaction,
+  TextChannel,
+  ThreadChannel,
+  User,
+} from 'discord.js'
+import {
+  bettingEmojis,
+  getAnyHandledEmojiKey,
+  getBetEmojiKey,
+  getBettingEmojisAsStrings,
+} from './emojis.js'
+import { registerHelpMessage, userApiKey } from './storage.js'
+
+const discordThreads: { [key: string]: ThreadChannel } = {}
+
+export const handleReaction = async (
+  reaction: MessageReaction,
+  user: User,
+  channel: TextChannel,
+  market: FullMarket
+) => {
+  const { name } = reaction.emoji
+  console.log(`Collected ${name} from user id: ${user.id}`)
+  if (!getAnyHandledEmojiKey(reaction)) {
+    console.log('Not a handled emoji')
+    return
+  }
+  // Market description
+  if (name === 'ℹ️') {
+    const content = `Market details: ${market.textDescription}`
+    await sendThreadMessage(channel, market, content, user, true)
+    return
+  }
+
+  // Help
+  if (name === '❓') {
+    const { yesBetsEmojis, noBetsEmojis } = getBettingEmojisAsStrings(
+      channel.guild
+    )
+    const content = `This is a market for the question "${market.question}". You can bet YES by reacting with these emojis: ${yesBetsEmojis} and NO with these: ${noBetsEmojis}. The emoji numbers correspond to the amount of mana used per bet.`
+    await user.send(content)
+    return
+  }
+  // The embeds don't load unless we fetch the message every time even though the message is not marked as partial
+  // seems related to: https://github.com/discordjs/discord.js/issues/7697#issuecomment-1073432737
+  const message =
+    reaction.message.embeds.length === 0 || reaction.message.partial
+      ? await reaction.message.fetch()
+      : reaction.message
+
+  const closed = (market.closeTime ?? 0) <= Date.now()
+  if (closed) {
+    await sendThreadMessage(
+      channel,
+      market,
+      getCurrentMarketDescription(market),
+      user
+    )
+    await updateMarketStatus(message, market)
+    return
+  }
+  // Attempt to place a bet
+  await handleBet(reaction, user, channel, message, market)
+}
+
+export const handleBet = async (
+  reaction: MessageReaction,
+  user: User,
+  channel: TextChannel,
+  message: Message,
+  market: FullMarket,
+  sale?: boolean
+) => {
+  const emojiKey = getBetEmojiKey(reaction)
+  if (!emojiKey) return
+  const { amount, outcome: buyOutcome } = bettingEmojis[emojiKey]
+  console.log('betting', amount, buyOutcome, 'on', market.id, 'for', user.tag)
+  try {
+    const apiKey = await userApiKey(user.id)
+    if (!apiKey) {
+      if (sale) return
+      await user.send(registerHelpMessage)
+      const userReactions = message.reactions.cache.filter((reaction) =>
+        reaction.users.cache.has(user.id)
+      )
+      try {
+        for (const reaction of userReactions.values()) {
+          await reaction.users.remove(user.id)
+        }
```

Comment:
Kind of weird that this literally removes all their reacts, not just the one they just made.

Reply:
good point! Fixed